signer/src/block_observer.rs

```diff
@@ -0,0 +1,425 @@
+//! # Block observer
+//!
+//! This module contains the block observer implementation for the sBTC signer.
+//! The block observer is responsible for populating the signer database with
+//! information from the Bitcoin and Stacks blockchains, and notifying
+//! the signer event loop whenever the state has been updated.
+//!
+//! The following information is extracted by the block observer:
+//! - Bitcoin blocks
+//! - Stacks blocks
+//! - Deposit requests
+//! - sBTC transactions
+//! - Withdraw requests
+//! - Deposit accept transactions
+//! - Withdraw accept transactions
+//! - Withdraw reject transactions
+//! - Update signer set transactions
+//! - Set aggregate key transactions
+
+use std::collections::HashMap;
+
+use crate::storage;
+
+use bitcoin::hashes::Hash;
+use blockstack_lib::chainstate::nakamoto;
+use blockstack_lib::chainstate::stacks;
+use futures::stream::StreamExt;
+use storage::model;
+use storage::Read;
+use storage::Write;
+
+type DepositRequestMap = HashMap<bitcoin::Txid, Vec<DepositRequest>>;
+
+/// Block observer
+pub struct BlockObserver<BitcoinClient, StacksClient, EmilyClient, BlockHashStream, SignerStorage> {
+    /// Bitcoin client
+    pub bitcoin_client: BitcoinClient,
+    /// Stacks client
+    pub stacks_client: StacksClient,
+    /// Emily client
+    pub emily_client: EmilyClient,
+    /// Stream of blocks from the block notifier
+    pub bitcoin_blocks: BlockHashStream,
+    /// Database connection
+    pub storage: SignerStorage,
+    /// Used to notify any other system that the database has been updated
+    pub subscribers: tokio::sync::watch::Sender<()>,
```

Comment:
This is fine, but we may want to add some other process/logic in the future that is more resilient to signer restarts or failures. A design that uses the databases state itself to helps subscribers know if they have work to do on boot by reading a table (or something like that).

This is probably a more general comment where, for v2, we'll want to go back and audit the codebase for cases where we'll want more resilience and add design docs for the changes we'll want to make.

Reply:
So I think this is pretty resilient, since both the block observer and signer event loop runs in the same binary (which we might not have explicitly said yet, but I see them as two tasks running within the same process), this simple channel is enough to guarantee that the signer event loop will know whenever the database has updated and it should fetch the latest information and do it's thing.

Reply:
Hmmm, yeah I think I need to be more clear about the problem that I was thinking about. If the sender notifies the subscriber about an event and we restart the binary while the subscriber is doing it's thing, will the respawned subscriber know that it still has work to do? I thought the answer was "no", but I'd be happy if I was wrong about that. Also, I think this will be resilient depending on the process we have for getting notified about new blocks.

But yeah I don't think there is a problem with this particular piece, especially right now. I was thinking a little more generally about ensuring that each component doesn't lose track of things when there is a failure or a restart.

Reply:
Yeah all of this relates to how we design the signer event loop. The way I picture it is that the signer event loop will read everything it needs from the database and process any pending requests it sees, based on the state. The signal from the block observer is essentially just a wakeup call saying "Hey! There's new data, please do your thing". On startup I'd imagine the signer event loop to always check the database and process any pending requests before starting to listen to the block observer.